### DISCUSSION TITLE
Connection error on triggering automation

### DISCUSSION THREAD

**Discussion by jovana** [WillCodeForCats/solaredge-modbus-multi#649] [Q&A]
I have run multiple automations that change the settings for the inverter.
In some cases, it needs to set 4 selections to have the correct configuration:

<img width="786" alt="Scherm­afbeelding 2024-08-08 om 12 21 47" src="https://github.com/user-attachments/assets/30324701-780c-4085-86dd-f04545dd5b22">

If the automation triggers changing the settings the inverter gives the error: "Error: Connection to inverter ID 2 failed".
<img width="550" alt="Scherm­afbeelding 2024-08-08 om 12 22 08" src="https://github.com/user-attachments/assets/f5d39bc7-437a-43dc-8201-62aca68fc5dd">

Now I have set 10 seconds in the "Inverter command delay" option. 
<img width="415" alt="Scherm­afbeelding 2024-08-09 om 11 28 14" src="https://github.com/user-attachments/assets/b78de833-5b2b-45b0-81fc-c978a857770f">

But it does not seem to help, because the automation is not always fully triggered.
Now have added a 15-second delay between each action, this seems to work.
<img width="737" alt="Scherm­afbeelding 2024-08-09 om 11 24 22" src="https://github.com/user-attachments/assets/85fea0d4-0292-418c-8d67-978ec7d05111">

Not a real issue, but the automation takes a long time to complete in this setup.

Quest: how can I solve this in a better and more cleaner way? Do you have any suggestions how to solve this?

Many thanks!
Jovana

**Comment by deqw** [comment]
Can't see what you are doing in your automations, but anyway don't change Storage Control Mode each time. Set and leave to Remote Control once and do your automations.

**Reply by jovana** [reply]
Thanks for your help!

Do you mean this option "Storage Control Mode" leave it to "Remote Control"?
<img width="363" alt="Scherm­afbeelding 2024-08-09 om 11 53 11" src="https://github.com/user-attachments/assets/b247a525-a7dd-4896-89ae-b91f2c19b176">

So I don't have to change this each time? I want to switch the battery charging the moment the price is low or the voltage is to high on the grid, but also fully discharge when the prices are high. And between running my house on its battery power.

I will change the automation and give it a try.

**Reply by deqw** [reply]
Yes that is it. I see I have a one second delay between each message to the inverter. What you are trying I have been doing for some time now (not with the HA Automations but with Node-Red). Reducing power when voltage is hight etc. But I have stopped all these automations and activated Time of Use in the SolarEdge app (they call it SolarEdge One) and let them do their smart things. Try it, it will save you a lot of time messing with the thing yourself.

**Reply by deqw** [reply]
BTW: When you activate Time of Use the first thing you see is that the Inverter will go into Remote Control.

**Reply by deqw** [reply]
Oh and when you have high Voltage (>253V) you will see your inverter stop producing to the grid like it normally does, but the after a few seconds it will charge your battery if there is room for it.

**Reply by jovana** [reply]
Many thanks for your help!

Yes I notice this option in the SE app also. But I was wondering if it also discharges to the grid when the price is high. In that case, HA automation is not needed anymore :-)

**Reply by jovana** [reply]
> BTW: When you activate Time of Use the first thing you see is that the Inverter will go into Remote Control.

It seems this option is not (jet) available for my inverter:
<img src="https://github.com/user-attachments/assets/7dee5669-f60b-4aa6-8110-5b1bdb6143aa" width="300" />

I have the SE5K-RWB48BFN4, it is probably too new for this feature on this moment.

**Reply by deqw** [reply]
I have the same, but the 10K version. What firmware are you on?

**Reply by jovana** [reply]
I have firmware: 0004.0020.0036

**Reply by deqw** [reply]
I am running 0004.0021.0019 at the moment, but that is still a beta version. But I know someone who has 4.20.36 and is using Time of Use. Your mySolarEdge app is up to date?

**Reply by deqw** [reply]
This: "Please note that the "Time of Use" feature is a beta version which is still undergoing final testing before its official release, is not yet available for all users and is offered on a pilot basis for select system owners."

Found that [here](https://marketing.solaredge.com/tou-mode).

**Reply by jovana** [reply]
Yes I have updated the app also. I think this feature is not release for this inverter. Need to wait for the next firmware update.
there is a release 4.21.103 already: https://www.solaredge.com/en/support/setapp-inverters-firmware

**Reply by deqw** [reply]
That one is only for the short string inverters. Just wait a bit or contact SE and ask to be included in the test already.

**Comment by deqw** [comment]
Yes it does that. I attach a screenshot. You see the discharging yesterday evening and it just started charging on solar. SE says there is some AI involved so let it do its thing for a while.

![Screenshot 2024-08-09 at 12 48 47](https://github.com/user-attachments/assets/06393bdd-c7bc-44e5-a8a3-edd42821ca51)

**Reply by jovana** [reply]
This is very nice!!
SE keeps working on improvements. A few months ago they didn't have this. I remember the launched in January to optimize for the dynamic energy plans.

I will keep an eye on this feature to release for my inverter also.

thanks again for the info!!!

**Reply by deqw** [reply]
What you also see in the graph is that it depleted the battery a bit too early last night. Still learning probably ;o)

**Reply by deqw** [reply]
The last nine days it worked actually quite well. You see it dumping into the grid when there is room. Yesterday is a bit off probably because I messed with it myself. What I did not see yet is charging from the grid eg. at night. But they know daytime solar will be sufficient for the day so charging at night will not be necessary. I also see my daily profits rising somewhat.
![Screenshot 2024-08-09 at 13 02 02](https://github.com/user-attachments/assets/3f1e9925-d8fa-4655-b66c-ff880faa4957)

**Reply by jovana** [reply]
Well 20 hour yesterday was a good moment to return the energy, the price was at the highest.

**Comment by jovana** [comment]
@WillCodeForCats Is it possible to have some sort of debounce mechanism for changing values for inverts? Or combinding changes in one command? 

I have two inverters (master / slave setup). And changing settings is almost any time an issue and get not go through. Basically, I change values for "Storage Command Mode", "Storage discharge limit" and "Storage Default Mode". I have now added delays for many minutes, but even then the inverters became unresponsive. The first command is always set, the second of third are not. Can we send one command containing the changes we need like:

In pseudo-code:
{
"storage_command_mode": option_a,
"storage_default_mode": option_b,
"storage_discharge_limit": 5000,
}

Probably there is also a way to send the info combined multiple values using yaml?
```
      - device_id: 281e1152847b69adcc6bd792904ac578
        domain: select
        entity_id: 705d87b43ec72ba5ba74b2cc724b0918
        type: select_option
        option: Maximize Self Consumption

      - device_id: 281e1152847b69adcc6bd792904ac578
        domain: number
        entity_id: f3b21bd8dcee5703ef14d0e99a563bbd
        type: set_value
        value: 0

      - device_id: 281e1152847b69adcc6bd792904ac578
        domain: select
        entity_id: 1e87c6f44ed6c6ff21632286dbdf9f0b
        type: select_option
        option: Maximize Self Consumption
```

**Reply by deqw** [reply]
I commented on this before: don't change Storage Control Mode each time; leave it to Remote Control. One second between consecutive commands for the other settings will be sufficient.

**Reply by jovana** [reply]
Yes I have followed your advice and have remove this from the automation. But I still need to change:
"storage_command_mode"
"storage_default_mode"
"storage_discharge_limit"

(I have edited my previous post, because I mixed you the storage_control_mode and storage_default_mode).

Thanks

**Reply by WillCodeForCats** [reply]
Modbus only accepts one register at a time, there is no group or bulk commands in modbus.

**Reply by deqw** [reply]
Not quite clear what you try to achieve, but why change the Storage Default Mode (each time) in an automation? You set this once and then  retrieve later to set the Storage Command Mode accordingly.

**Reply by jovana** [reply]
> Modbus only accepts one register at a time, there is no group or bulk commands in modbus.

Okay thanks!

**Reply by jovana** [reply]
> Not quite clear what you try to achieve, but why change the Storage Default Mode (each time) in automation? You set this once and then retrieve later to set the Storage Command Mode accordingly.

Well not sure, but if I understand correctly I need to change both settings: "storage_command_mode" and "storage_default_mode", right? Only change one is not giving the desired results.

Okay, what I want to do is this. I have triggers based on electricity pricing, solar power, charging car, grid voltage monitor etc. Here are a few examples:
- Once a day (max one hour) discharge to maximize export (only if the difference between low and high prices is more than 12 ct).
- If there is no solar power: Discharge to minimize export
- If there is too much solar power and the voltage on the grid is over 252volts: charge from solar power (otherwise the inverter will shuts down)
- If the price is on the lowest hours (max 2 a day) Charge from Solar Power and grid.
- etc. 

In this image, the 3 settings I have highlighted I need to change to achieve this right?
<img width="200" alt="Scherm­afbeelding 2024-08-31 om 09 51 57" src="https://github.com/user-attachments/assets/cc69f507-dd6e-4308-bb93-2f9e935cf091">

**Reply by deqw** [reply]
I never needed to change the 'Storage Default Mode' in an automation or programmatically to change these other settings. I would say: don't touch it and leave it to 'Maximize Self Consumption'.

- If there is no solar power: => I would choose 'Maximize Self Consumption' here

When putting your examples in different automations I can see these automations easily working against each other. Eg. what if voltage is over 252V and the price is lowest of the day? You don't want to maintain this kind of automations, I would say.

I have been programming around this myself for a few months in Home Assistant / Node Red). There are so many circumstances/variables that you have to catch that in the end it isn't worth your time (at least not mine ;o). Like I mentioned before I'm now 100% on SolarEdge One and use the integration for reading only.

What is your inverter firmware version today?

**Reply by jovana** [reply]
Thanks for the advice (again)!
For now, I leave the storage default mode to MSC. Let's see what happened and keep you posted.

I understand very well that conflicts can occur because of different triggers. 
To prevent this, I have only one automation only that can change settings to the inverter. Via a helper select option box other automation can change this helper. Each automation will run 1 minute past every hour (because on price change trigger I have to run my Python script for calculation stuff).

Because I know sending too many commands makes the inverter less responsive I try to avoid this.

**Reply by deqw** [reply]
Do you push "button.solaredge_i1_commit_power_settings" somewhere in your automations? This command typically takes 5 to 10 seconds for the inverter and is not necessary for what you try to accomplish. So avoid that too.

**Reply by jovana** [reply]
No, I do not use this commit option. I did not understand why this button exists if we can not setup registers first and then commit settings.

**Reply by WillCodeForCats** [reply]
The documentation says to use the commit for some settings, such as changing Reactive Power Mode.

**Reply by WillCodeForCats** [reply]
The docs also say Current Limit and Power Reduce also fall under settings that use commit.

These are different from Active Power Limit which is documented as dynamic.

The difference between dynamic and "committed" settings the documentation says dynamic commands are not saved when the inverter restarts.

However, none of my inverters accept commands so I can't personally confirm this.

**Reply by Mougli** [reply]
@jovana I was encountering the same issue trying to do something similar to you.

I have access to the Time of Use mode, but it doesn't let you set a reduced rate window, so it doesn't charge when energy is the cheapest so I had to go manual instead.

In the end, what's fixed it for me is to tick the "Keep Modbus Connection Open" box in the integration's configuration (I also set the inverter command delay to 10 seconds, but haven't checked if that's really necessary).

Then all my automation need to do is change the storage command mode (no refresh or commit required).

Note that you might want to adjust the storage command timeout, as after that duration, the inverter will revert to whatever mode is set in Storage Default Mode.

So if you want to charge for an hour max, your one time setup would be:

- Storage Control Mode: Remote Control
- Storage Command timeout: 3600s
- Storage Default Mode: Maximise self consumption.

Then you just need an automation that changes the Storage Command Mode to "Charge from Solar and Grid", and when it triggers, it will revert back to MSC on its own after an hour.
